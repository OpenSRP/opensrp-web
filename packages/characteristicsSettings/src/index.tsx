export * from './components/EditSettings';
export * from './components/LocationsMenu';
export * from './components/SearchForm';
export * from './ducks/locations';
export * from './ducks/settings';
