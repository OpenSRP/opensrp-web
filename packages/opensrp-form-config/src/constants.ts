export const MAKE_RELEASE_LABEL = 'Make Release';
export const FILE_NAME_LABEL = 'File Name';
export const FILE_VERSION_LABEL = 'File Version';
export const IDENTIFIER_LABEL = 'Identifier';
export const MODULE_LABEL = 'Module';
export const EDIT_LABEL = 'Edit';
export const DOWNLOAD_LABEL = 'Download';
export const UPLOAD_EDIT_LABEL = 'Upload edit';
