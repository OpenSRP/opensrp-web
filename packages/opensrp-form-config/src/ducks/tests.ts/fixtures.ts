export const fixManifestReleases = [
    {
        identifier: '1.0.11',
        json: '{"forms_version":"1.0.11","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.14',
        createdAt: '2020-06-08T15:00:38.413+02:00',
        updatedAt: '2020-06-08T15:00:38.413+02:00',
    },
    {
        identifier: '1.0.12',
        json: '{"forms_version":"1.0.12","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.14',
        createdAt: '2020-06-08T15:03:00.950+02:00',
        updatedAt: '2020-06-08T15:03:00.950+02:00',
    },
    {
        identifier: '1.0.16',
        json: '{"forms_version":"1.0.16","identifiers":["blood_screening.json"}',
        appId: 'org.smartregister.reveal',
        appVersion: '1.2.16',
        createdAt: '2020-06-09T14:40:04.189+02:00',
        updatedAt: '2020-06-09T14:40:04.189+02:00',
    },
];
