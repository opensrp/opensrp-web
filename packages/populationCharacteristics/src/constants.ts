export const SEARCH_LABEL = 'Search';
export const SEARCH_SETTINGS_LABEL = 'Search Settings';
export const PAGE_TITLE_LABEL = 'Server Settings';
export const EDIT_LABEL = 'Edit';
export const DESCRIPTION_LABEL = 'Description';
export const SETTINGS_LABEL = 'Settings';
export const INHERITED_FROM_LABEL = 'Inherited from';
export const NAME_LABEL = 'Name';
export const SET_TO_YES_LABEL = `Set to 'Yes'`;
export const SET_TO_NO_LABEL = `Set to 'No'`;
export const INHERIT_SETTING_LABEL = 'Inherit setting';
export const NO_DATA_FOUND = 'No data found';
