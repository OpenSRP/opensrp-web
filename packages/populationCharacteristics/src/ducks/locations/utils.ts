/**
 * Given the location id and the loc map, return the location
 * details by traversing the loc map hierarchy
 * @param locMap - the loc map
 * @param locId - the loc id
 */

import { LocChildren, LocMap } from '.';

export function getLocationDetailsFromLocMap(locMap: LocMap, locId: string): LocChildren {
    const locationIds = Object.keys(locMap);
    let i = 0;
    let locFound = false;
    let locDetails = null;

    while (!locFound && i < locationIds.length) {
        const currentLocationId = locationIds[i];
        const currentLocDetails = locMap[currentLocationId];

        if (currentLocationId === locId) {
            locDetails = currentLocDetails;
            locFound = true;
        } else {
            if (currentLocDetails.children) {
                return getLocationDetailsFromLocMap(currentLocDetails.children, locId);
            }
        }

        i += 1;
    }

    return locDetails || ({} as LocChildren);
}
