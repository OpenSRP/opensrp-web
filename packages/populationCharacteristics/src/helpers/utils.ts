import { Setting } from '../ducks/settings';
import {
    DESCRIPTION_LABEL,
    EDIT_LABEL,
    INHERITED_FROM_LABEL,
    INHERIT_SETTING_LABEL,
    NAME_LABEL,
    NO_DATA_FOUND,
    PAGE_TITLE_LABEL,
    SEARCH_SETTINGS_LABEL,
    SETTINGS_LABEL,
    SET_TO_NO_LABEL,
    SET_TO_YES_LABEL,
} from '../constants';
import { EditSettingLabels } from './types';

interface PostData extends Partial<Setting> {
    _id: string;
    identifier: string;
}

/**
 *
 * @param {Setting} data
 * @param {boolean} value
 */
export const preparePutData = (data: Setting, value: string) => {
    const {
        description,
        key,
        label,
        locationId,
        providerId,
        uuid,
        settingIdentifier: identifier,
        settingMetadataId: _id,
        team,
        teamId,
        type,
        documentId,
    } = data;
    let postData: PostData = {
        _id,
        description,
        identifier,
        key,
        label,
        locationId,
        uuid,
        settingsId: documentId,
        type,
        value,
    };
    if (team) postData = { ...postData, team };
    if (teamId) postData = { ...postData, teamId };
    if (providerId) postData = { ...postData, providerId };
    return postData;
};

export const labels: EditSettingLabels = {
    descriptionLabel: DESCRIPTION_LABEL,
    editLabel: EDIT_LABEL,
    inheritedLable: INHERITED_FROM_LABEL,
    inheritSettingsLabel: INHERIT_SETTING_LABEL,
    nameLabel: NAME_LABEL,
    noDataFound: NO_DATA_FOUND,
    pageTitle: PAGE_TITLE_LABEL,
    placeholder: SEARCH_SETTINGS_LABEL,
    settingLabel: SETTINGS_LABEL,
    setToNoLabel: SET_TO_NO_LABEL,
    setToYesLabel: SET_TO_YES_LABEL,
};
