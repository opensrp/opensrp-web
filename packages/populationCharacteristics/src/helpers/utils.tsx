import React, { MouseEvent, useEffect, useRef, cloneElement, ReactElement } from 'react';
import { Setting } from '../ducks/settings';
import {
    DESCRIPTION_LABEL,
    EDIT_LABEL,
    INHERITED_FROM_LABEL,
    INHERIT_SETTING_LABEL,
    NAME_LABEL,
    NO_DATA_FOUND,
    PAGE_TITLE_LABEL,
    SEARCH_SETTINGS_LABEL,
    SETTINGS_LABEL,
    SET_TO_NO_LABEL,
    SET_TO_YES_LABEL,
    ACTION_LABEL,
    SETTINGS_INHERIT,
    SETTINGS_FALSE,
    SETTINGS_TRUE,
    TOOLTIP_INHERITED_FROM,
} from '../constants';
import { EditSettingLabels, SettingValue } from './types';

interface PostData extends Partial<Setting> {
    _id: string;
    identifier: string;
}

/**
 *
 * @param {Setting} data
 * @param {boolean} value
 */
export const preparePutData = (data: Setting, value: string) => {
    const {
        description,
        key,
        label,
        locationId,
        providerId,
        uuid,
        settingIdentifier: identifier,
        settingMetadataId: _id,
        team,
        teamId,
        type,
        documentId,
    } = data;
    let postData: PostData = {
        _id,
        description,
        identifier,
        key,
        label,
        locationId,
        uuid,
        settingsId: documentId,
        type,
        value,
    };
    if (team) postData = { ...postData, team };
    if (teamId) postData = { ...postData, teamId };
    if (providerId) postData = { ...postData, providerId };
    return postData;
};

export const labels: EditSettingLabels = {
    actionLabel: ACTION_LABEL,
    descriptionLabel: DESCRIPTION_LABEL,
    editLabel: EDIT_LABEL,
    inheritedLable: INHERITED_FROM_LABEL,
    inheritSettingsLabel: INHERIT_SETTING_LABEL,
    nameLabel: NAME_LABEL,
    noDataFound: NO_DATA_FOUND,
    pageTitle: PAGE_TITLE_LABEL,
    placeholder: SEARCH_SETTINGS_LABEL,
    settingLabel: SETTINGS_LABEL,
    setToNoLabel: SET_TO_NO_LABEL,
    setToYesLabel: SET_TO_YES_LABEL,
    toolTipInheritedFrom: TOOLTIP_INHERITED_FROM,
};

/** format table data function props interface */
interface EditSettingsButtonProps {
    changeSetting: (event: MouseEvent<HTMLDivElement>, setting: Setting, value: SettingValue) => void;
    editLabel: string;
    inheritSettingsLabel: string;
    openEditModal: (event: MouseEvent<HTMLAnchorElement>, setting: Setting) => void;
    row: Setting;
    setToNoLabel: string;
    setToYesLabel: string;
    value: boolean;
    showInheritSettingsLabel: boolean;
}

interface ClickOutsideProps {
    children: ReactElement;
    onClick: any;
}

export const ClickOutside = ({ children, onClick }: ClickOutsideProps) => {
    const ref = useRef<HTMLDivElement>(null);

    useEffect(() => {
        const handleClickOutside = (e: any) => {
            if (ref.current && !ref.current.contains(e.target)) {
                onClick(e);
            }
        };
        document.addEventListener('mousedown', handleClickOutside);
        return () => {
            document.removeEventListener('mousedown', handleClickOutside);
        };
    }, [onClick]);
    return <>{cloneElement(children, { ref })}</>;
};

export const EditSettingsButton = (props: EditSettingsButtonProps) => {
    const {
        editLabel,
        inheritSettingsLabel,
        value,
        setToYesLabel,
        setToNoLabel,
        row,
        openEditModal,
        changeSetting,
        showInheritSettingsLabel,
    } = props;

    const onClickOutSide = React.useCallback(
        (e: any) => {
            if (row.editing) {
                openEditModal(e, row);
            }
        },
        [row.editing],
    );

    return (
        <div className="popup" key={row.key}>
            <a href="#" onClick={e => openEditModal(e, row)}>
                {editLabel}
            </a>
            <ClickOutside onClick={onClickOutSide}>
                <div className={`popuptext ${row.editing ? 'show' : ''}`}>
                    <div onClick={e => changeSetting(e, row, SETTINGS_TRUE)}>
                        <span className={value && !row.inheritedFrom ? 'check' : 'empty-check'} />
                        <span>{setToYesLabel}</span>
                    </div>
                    <div onClick={e => changeSetting(e, row, SETTINGS_FALSE)}>
                        <span className={value || row.inheritedFrom ? 'empty-check' : 'check'} />
                        <span>{setToNoLabel}</span>
                    </div>
                    {showInheritSettingsLabel && (
                        <div onClick={e => changeSetting(e, row, SETTINGS_INHERIT)} className="inherit-from">
                            <span className={row.inheritedFrom?.trim() ? 'check' : 'empty-check'} />
                            <span>{inheritSettingsLabel}</span>
                        </div>
                    )}
                </div>
            </ClickOutside>
        </div>
    );
};
