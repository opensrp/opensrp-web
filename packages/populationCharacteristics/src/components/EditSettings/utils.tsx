import { SETTINGS_INHERIT } from '../../constants';
import { Setting, fetchLocSettings, getLocSettings } from '../../ducks/settings';
import { SettingValue } from '../../helpers/types';
import { getLocDetails } from '../../ducks/locations';
import { Store } from 'redux';
import { OpenSRPService } from '@opensrp/server-service';
import { getFetchOptions } from '@opensrp/server-service';
import { ToastOptions } from 'react-toastify';
import { preparePutData } from '../..//helpers/utils';

/**
 * Update redux store after a successful response from the API
 * if a setting is edited
 * @param row - the setting object
 * @param value - the selected value for the setting
 * @param state - the redux store
 * @param fetchSettings - the redux action
 * @param activeLocationId - the active location id
 */
export const onEditSuccess = (
    state: Partial<Store>,
    row: Setting,
    value: SettingValue,
    fetchSettings: typeof fetchLocSettings,
    activeLocationId: string,
) => {
    if (value !== SETTINGS_INHERIT) {
        // We set the new value and make sure to override inheritedFrom
        // to none
        fetchSettings([{ ...row, editing: false, value, inheritedFrom: '' }], activeLocationId);
    } else {
        // Get the parent of active location and get the parent setting matching
        // the setting we are editing and inherit its value
        const locationDetails = getLocDetails(state, activeLocationId);
        const parentId = locationDetails.node.parentLocation?.locationId;

        if (parentId) {
            const parentSettings = getLocSettings(state, parentId);

            if (parentSettings.length) {
                let inheritedValue = row.value;
                let settingFound = false;
                let i = 0;

                while (!settingFound && i < parentSettings.length) {
                    const currentSetting = parentSettings[i];

                    if (currentSetting.key === row.key) {
                        inheritedValue = currentSetting.value;
                        settingFound = true;
                    }

                    i += 1;
                }

                if (settingFound) {
                    fetchSettings(
                        [{ ...row, editing: false, value: inheritedValue, inheritedFrom: parentId }],
                        activeLocationId,
                    );
                }
            }
        }
    }
};

/**
 * Update setting by calling the OpenSRP endpoint and updating the store if the update is
 * a success
 * @param state - redux store
 * @param row - the setting object
 * @param value - the new value
 * @param v2BaseUrl - OpenSRP base URL
 * @param settingsEndpoint - the endpoint
 * @param getPayload - options that are to be added to openSRPService fetch request
 * @param fetchSettings - redux action
 * @param activeLocationId - active location id
 * @param customAlert - a custom toast alert
 */
export const editSetting = async (
    state: Partial<Store>,
    row: Setting,
    value: SettingValue,
    v2BaseUrl: string,
    settingsEndpoint: string,
    getPayload: typeof getFetchOptions,
    fetchSettings: typeof fetchLocSettings,
    activeLocationId: string,
    customAlert?: (message: string, options: ToastOptions) => void,
) => {
    if ((value === SETTINGS_INHERIT && row.inheritedFrom) || (value === row.value && !row.inheritedFrom)) {
        return;
    }

    const endPoint =
        value === SETTINGS_INHERIT || activeLocationId === row.locationId
            ? `${settingsEndpoint}${row.settingMetadataId}`
            : settingsEndpoint;

    if (value === SETTINGS_INHERIT) {
        const clientService = new OpenSRPService(v2BaseUrl, endPoint, getPayload);

        await clientService
            .delete()
            .then(() => {
                onEditSuccess(state, row, value, fetchSettings, activeLocationId);
            })
            .catch(error => {
                customAlert && customAlert(String(error), { type: 'error' });
            });
    } else {
        const data = preparePutData(row, value);

        if (activeLocationId !== row.locationId) {
            data.locationId = activeLocationId;
            delete data.uuid;
            delete data._id;
            const clientService = new OpenSRPService(v2BaseUrl, endPoint, getPayload);
            await clientService
                .create(data)
                .then(() => {
                    onEditSuccess(state, row, value, fetchSettings, activeLocationId);
                })
                .catch(error => {
                    customAlert && customAlert(String(error), { type: 'error' });
                });
        } else {
            const clientService = new OpenSRPService(v2BaseUrl, endPoint, getPayload);
            await clientService
                .update(data)
                .then(() => {
                    onEditSuccess(state, row, value, fetchSettings, activeLocationId);
                })
                .catch(error => {
                    customAlert && customAlert(String(error), { type: 'error' });
                });
        }
    }
};

/**
 * Checks if the inherited from value is a valid location in the location hierarchy
 * @param state - redux store
 * @param inheritedFrom - the location id the setting inherits from
 */
export const isInheritedFromValid = (state: Partial<Store>, inheritedFrom: string) => {
    const label = getLocDetails(state, inheritedFrom).label;

    if (!label) {
        return false;
    }

    return true;
};

/**
 * Returns the label to display when provided with the inherited from location id
 * @param state - redux store
 * @param inheritedFrom - the location id the setting inherits from
 */

export const getInheritedFromLabel = (state: Partial<Store>, inheritedFrom: string | undefined): string => {
    if (inheritedFrom) {
        const label = getLocDetails(state, inheritedFrom).label;

        if (label) {
            inheritedFrom = label;
        }
    } else {
        inheritedFrom = '_';
    }

    return inheritedFrom;
};
