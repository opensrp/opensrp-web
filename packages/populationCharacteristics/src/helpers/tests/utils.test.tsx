import React from 'react';
import { mount, shallow } from 'enzyme';
import { preparePutData, EditSettingsButton } from '../utils';
import { allSettings, setting3, setting4 } from '../../ducks/settings/tests/fixtures';
import { updateDate } from '../../components/EditSettings/tests/fixtures';
import { EDIT_LABEL } from '../../../../../clients/core/src/constants';
import { INHERIT_SETTING_LABEL, SET_TO_NO_LABEL, SET_TO_YES_LABEL } from '../../constants';
import toJson from 'enzyme-to-json';

const data = { ...allSettings[0] };
const output = {
    ...updateDate,
};

const mockFn = jest.fn();

describe('helpers/utils: preparePutData', () => {
    it('preparePutData', () => {
        const teamData = {
            team: 'demo',
            teamId: '123',
        };
        expect(preparePutData(data, 'true')).toEqual(output);

        // with teams
        expect(preparePutData({ ...data, ...teamData }, 'true')).toEqual({ ...output, ...teamData });

        // with providerId
        expect(preparePutData({ ...data, providerId: '12' }, 'true')).toEqual({ ...output, providerId: '12' });
    });
});

describe('helpers/utils: EditSettingsButton', () => {
    const editSettingsButtonProps = {
        changeSetting: mockFn,
        editLabel: EDIT_LABEL,
        inheritSettingsLabel: INHERIT_SETTING_LABEL,
        openEditModal: mockFn,
        row: allSettings[0],
        setToNoLabel: SET_TO_NO_LABEL,
        setToYesLabel: SET_TO_YES_LABEL,
        value: false,
        showInheritSettingsLabel: true,
    };

    afterEach(() => {
        jest.clearAllMocks();
        jest.resetAllMocks();
        jest.restoreAllMocks();
    });

    it('renders without crashing', () => {
        shallow(<EditSettingsButton {...editSettingsButtonProps} />);
    });

    it('renders correctly', () => {
        const wrapper = shallow(<EditSettingsButton {...editSettingsButtonProps} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('handles click Set to Yes', () => {
        const wrapper = mount(<EditSettingsButton {...editSettingsButtonProps} />);
        const editButton = wrapper.find('.popup a').at(0);
        editButton.simulate('click');
        wrapper.update();
        wrapper
            .find('.popuptext div')
            .at(0)
            .simulate('click');
        expect(editSettingsButtonProps.changeSetting).toHaveBeenCalledWith(
            expect.any(Object),
            editSettingsButtonProps.row,
            'true',
        );
    });

    it('handles click Set to No', () => {
        const wrapper = mount(<EditSettingsButton {...editSettingsButtonProps} />);
        const editButton = wrapper.find('.popup a').at(0);
        editButton.simulate('click');
        wrapper.update();
        wrapper
            .find('.popuptext div')
            .at(1)
            .simulate('click');
        expect(editSettingsButtonProps.changeSetting).toHaveBeenCalledWith(
            expect.any(Object),
            editSettingsButtonProps.row,
            'false',
        );
    });

    it('hanldes click Inherit Setting', () => {
        const wrapper = mount(<EditSettingsButton {...editSettingsButtonProps} />);
        const editButton = wrapper.find('.popup a').at(0);
        editButton.simulate('click');
        wrapper.update();
        wrapper
            .find('.popuptext div')
            .at(2)
            .simulate('click');
        expect(editSettingsButtonProps.changeSetting).toHaveBeenCalledWith(
            expect.any(Object),
            editSettingsButtonProps.row,
            'inherit',
        );
    });

    it('renders correctly when value is true', () => {
        const props = {
            ...editSettingsButtonProps,
            row: allSettings[1],
            value: true,
        };
        const wrapper = mount(<EditSettingsButton {...props} />);

        // edit button works correctly
        const editButton = wrapper.find('.popup a').at(0);
        editButton.simulate('click');
        wrapper.update();
        // true is checked
        expect(
            wrapper
                .find('.popuptext div')
                .at(0)
                .find('.check').length,
        ).toEqual(1);
        // false is not checked
        expect(
            wrapper
                .find('.popuptext div')
                .at(1)
                .find('.check').length,
        ).toEqual(0);

        // inherit is not checked
        expect(
            wrapper
                .find('.popuptext div')
                .at(2)
                .find('.check').length,
        ).toEqual(0);
    });

    it('renders correcly when showInheritSettingsLabel is false', () => {
        const props = {
            ...editSettingsButtonProps,
            showInheritSettingsLabel: false,
        };
        const wrapper = mount(<EditSettingsButton {...props} />);
        expect(wrapper.find('.popuptext div').length).toBe(2);
        expect(
            wrapper
                .find('.popuptext div')
                .at(0)
                .find('span')
                .at(1)
                .text(),
        ).toEqual("Set to 'Yes'");
        expect(
            wrapper
                .find('.popuptext div')
                .at(1)
                .find('span')
                .at(1)
                .text(),
        ).toEqual("Set to 'No'");
    });

    it('renders correctly for inherited setting', () => {
        const props = {
            ...editSettingsButtonProps,
            row: setting3,
        };
        const wrapper = mount(<EditSettingsButton {...props} />);
        // true is not checked
        expect(
            wrapper
                .find('.popuptext div')
                .at(0)
                .find('.check').length,
        ).toEqual(0);
        // false is not checked
        expect(
            wrapper
                .find('.popuptext div')
                .at(1)
                .find('.check').length,
        ).toEqual(0);
        // inherit is checked
        expect(
            wrapper
                .find('.popuptext div')
                .at(2)
                .find('.check').length,
        ).toEqual(1);
    });

    it('renders correctly if setting editing is true', () => {
        const props = {
            ...editSettingsButtonProps,
            row: setting4,
        };
        const wrapper = mount(<EditSettingsButton {...props} />);
        expect(
            wrapper
                .find('div')
                .at(1)
                .prop('className'),
        ).toEqual('popuptext show');
    });
});
