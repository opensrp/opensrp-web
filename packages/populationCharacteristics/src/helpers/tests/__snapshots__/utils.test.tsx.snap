// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers/utils: EditSettingsButton renders correctly 1`] = `
<div
  className="popup"
  key="pop_undernourish"
>
  <a
    href="#"
    onClick={[Function]}
  >
    Edit
  </a>
  <ClickOutside
    onClick={[Function]}
  >
    <div
      className="popuptext "
    >
      <div
        onClick={[Function]}
      >
        <span
          className="empty-check"
        />
        <span>
          Set to 'Yes'
        </span>
      </div>
      <div
        onClick={[Function]}
      >
        <span
          className="check"
        />
        <span>
          Set to 'No'
        </span>
      </div>
      <div
        className="inherit-from"
        onClick={[Function]}
      >
        <span
          className="empty-check"
        />
        <span>
          Inherit setting
        </span>
      </div>
    </div>
  </ClickOutside>
</div>
`;
